Upgrade to latest Guava (v30) and Solr (v8) to resolve security alerts

tdonohue opened the pull request:
## References
This PR resolves 3 security alerts by upgrading Guava:
* https://github.com/DSpace/DSpace/network/alert/pom.xml/com.google.guava:guava/open
* https://github.com/DSpace/DSpace/network/alert/dspace/pom.xml/com.google.guava:guava/open
* https://github.com/DSpace/DSpace/network/alert/dspace-oai/pom.xml/com.google.guava:guava/open

This PR also moves us off an EOL version of Solr, resolving https://jira.lyrasis.org/browse/DS-4497

## Description
Upgrades the following dependencies:
* Upgrades `guava` from v19.0 to v30.0-jre. This resolves the security alerts listed above.
* Upgrades Solr from v7.3.1 to v8.7.0.  This was necessitated by the `guava` upgrade, as I was unable to get the latest version of `guava` to work with Solr v7 (because of Java API changes in guava).
* Minor upgrade of Jetty from v9.4.17.v20190418 to 9.4.35.v20201120 to align with the above upgrades and [fix GitHub security alert](https://github.com/DSpace/DSpace/security/dependabot/pom.xml/org.eclipse.jetty:jetty-server/open)
* Other minor dependency realignment in POMs to resolve dependency convergence issues.

Code changes required by Guava upgrade:
* Removes & replaces outdated/obsolete Lyncode `builder-commons` & `test-support` dependencies (from `dspace-oai` module classes).  These old Lyncode dependencies brought in a very old version of guava & were causing a headache in terms of dependency convergence.  *It was easiest to completely remove them & replace them with updated code that did the same thing.*
    * In order to determine what the Lyncode builders were doing, I looked at the source code at https://github.com/lyncode/builder-commons and found a modern replacement (in some cases using Guava directly & in others just using standard Java 11 code).
    * The Lyncode test-support classes were only used in `PipelineTest`, and I found replacement code in our existing [`QDCXslTest`](https://github.com/DSpace/DSpace/blob/main/dspace-oai/src/test/java/org/dspace/xoai/tests/stylesheets/QDCXslTest.java) class which does a similar test without using Lyncode's test-support. 
 
Code changes required by Solr upgrade:
* Updated `SolrSearchCore` and `SolrServiceImpl` to use `GET` requests when the `EmbeddedSolrServer` is in use (during integration test execution). 
    * This was necessary to work around a known issue with `POST` requests with `EmbeddedSolrServer`: https://issues.apache.org/jira/browse/SOLR-12858
* Updated `IndexFactoryImpl` to no longer use `ContentStreamUpdateRequest` for indexing full text.  Instead, it has been refactored to use Apache Tika directly.  Also introduced a new `discovery.solr.fulltext.charLimit` configuration to allow for more control over Tika's default character limit indexing settings.
    * This was necessary to work around recent changes/bugs introduced to `ContentStreamUpdateRequest` which are best described in this ticket https://issues.apache.org/jira/browse/SOLR-12798
    * Those bugs in `ContentStreamUpdateRequest` resulted in frequent errors during full text indexing (`o.e.j.h.HttpParser URI is too large >8192`) because the `ContentStreamUpdateRequest` class no longer supports multipart post well & was sending all fields as URL parameters.
    * Moving to using Apache Tika directly is also the recommended way to do full text indexing in Solr. See https://lucene.apache.org/solr/guide/8_7/uploading-data-with-solr-cell-using-apache-tika.html
* Minor updates to existing `solrconfig.xml` files to ensure they are aligned with Solr v8.7.0 default settings. (Some whitespace changes occurred here too that were automated by my IDE)
* Updated Docker to support Solr v8 (and the new directory structure in Solr v8 docker images).  Also required moving to using the official Solr Docker image with Docker Compose, instead of creating our own custom image.
    
Additional minor changes:
* Renamed `DatabaseRegistryUpdater` to `RegistryUpdater`.  This ensures it comes alphabetically *after* the `GroupServiceInitializer` and therefore is always run *after* Groups are created (as Flyway defaults to running Callbacks alphabetically).  I stumbled on this problem when starting with a fresh database...as the RegistryUpdater depends on the GroupServiceInitializer.

## Instructions for Reviewers
* Review code
* Verify tests still succeed (they do)
* Test OAI-PMH interface & reindexing.
* Test Solr indexing (`./dspace index-discovery -b`)
* Test search/browse functionality.

(NOTE: I've tested & verified all of the above and found no remaining issues.)

## Checklist

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

tdonohue committed:
Upgrade to latest guava. Remove/replace outdated Lyncode builders & test-support which used old guava version.

tdonohue commented:
Strangely, in this PR, the ITs are failing because of this error: #2989 (at least that's the main error logged in the test logs).   The changes in this PR are entirely unrelated to that issue, but perhaps that error was somehow being "swallowed" prior to the changes in this PR. Either that, or maybe an underlying error is being hidden/swallowed while only the ClassCastException is shown.

In any case, **this PR might be dependent on finding a fix for #2989.**

This PR works in manual testing. Server webapp works. OAI-PMH works, and `./dspace oai` commands work.  It's just the ITs which are having odd issues.

tdonohue committed:
Minor cleanup. Comment correction. Remove commented out code

tdonohue committed:
Rename to RegistryUpdater so that it comes alphabetically *after* GroupServiceInitializer as Flyway runs callbacks alphabetically

tdonohue committed:
Workaround https://issues.apache.org/jira/browse/SOLR-12858 by ensuring EmbeddedSolrServer always uses GET instead of POST

tdonohue committed:
Upgrade to Solr 8.7.0. Fix dependency convergence issues. Minor fix to MockSolrServer for Solr 8 compatibility

tdonohue committed:
Update Docker for Solr v8

tdonohue committed:
Update solrconfig.xml files for Solr v8 (also minor alignment correction)

tdonohue committed:
Fix/Workaround for https://issues.apache.org/jira/browse/SOLR-12798 . Use Tika directly for parsing instead of ContentStreamUpdateRequest (which results in "URI is too large >8192" errors in Solr v8)

tdonohue committed:
Add explicit dependency on Apache Tika to avoid UnsatisfiedDependencyException warnings from Spring Boot during startup

tdonohue commented:
This PR is now ready for review/testing.  I've scheduled it for Beta 5 in order to ensure it can be added prior to Testathon & 7.0 final.

tdonohue committed:
Docker bug fix. Ensure new Solr v8 data directories are kept between Docker restarts.

tdonohue commented:
@mwoodiupui : Would you be interested in giving this a review and/or test?  I'd appreciate your feedback here since you did the last Solr upgrade.  In addition, your feedback would be useful in the follow-up PR #3126 (where I've attempted to cleanup/remove old Solr configs that seem unused/unnecessary)

mwoodiupui commented on dspace-api/src/main/java/org/dspace/discovery/indexobject/IndexFactoryImpl.java:
```diff
@@ -74,32 +79,59 @@ public void writeDocument(Context context, T indexableObject, SolrInputDocument
+                    if (saxe.getMessage().contains("limit has been reached")) {
+                        // log that we only indexed up to that configured limit
+                        log.info("Full text is larger than the configured limit (discovery.solr.fulltext.charLimit)."
+                                     + " Only the first " + charLimit + " characters were indexed.");
```
Log4j2's parameter substitution would reduce this expensive multiple-string-concatenation expression to a compile-time constant.

mwoodiupui commented on dspace-api/src/main/java/org/dspace/discovery/indexobject/IndexFactoryImpl.java:
```diff
@@ -74,32 +79,59 @@ public void writeDocument(Context context, T indexableObject, SolrInputDocument
+
+                // Add document & commit immediately
+                solr.add(doc);
+                solr.commit(true, true);
```
Do we really want a hard commit after every document?  The usual advice from the Solr maintainers on committing in the client is "don't."  If this method is ever used in bulk indexing, the bulk operation will be quite slow *and* make the top-level caches ineffective for all other users.  We should tune our autocommit settings for a good balance of indexing performance and visibility, and trust autocommit.

tdonohue replied:
Good point.  To be honest, I debated on this one myself...I only added this in because the code it replaced was also doing a hard commit.  So, I did an exact replacement here.  That said, I'm OK with removing this in favor of autocommit.

mwoodiupui commented on dspace-api/src/test/java/org/dspace/solr/MockSolrServer.java:
```diff
@@ -159,9 +162,9 @@ public void destroy() throws Exception {
-            String solrDir = AbstractDSpaceIntegrationTest.getDspaceDir() + File.separator + "solr";
-            log.info("Initializing SOLR CoreContainer with directory " + solrDir);
-            container = new CoreContainer(solrDir);
+            Path solrDir = Paths.get(AbstractDSpaceIntegrationTest.getDspaceDir() + File.separator + "solr");
```
`Paths.get(AbstractDSpaceIntegrationTest.getDspaceDir(), "solr");`

mwoodiupui commented on dspace-api/src/test/java/org/dspace/solr/MockSolrServer.java:
```diff
@@ -159,9 +162,9 @@ public void destroy() throws Exception {
-            log.info("Initializing SOLR CoreContainer with directory " + solrDir);
-            container = new CoreContainer(solrDir);
+            Path solrDir = Paths.get(AbstractDSpaceIntegrationTest.getDspaceDir() + File.separator + "solr");
+            log.info("Initializing SOLR CoreContainer with directory " + solrDir.toAbsolutePath().toString());
```
Parameter substitution here.

tdonohue committed:
Minor refactors based on feedback

mwoodiupui commented on dspace-api/pom.xml:
```diff
@@ -524,133 +566,33 @@
         </dependency>
         <dependency>
             <groupId>org.apache.solr</groupId>
             <artifactId>solr-cell</artifactId>
```
We no longer have a direct dependency on solr-cell, but we've picked up a direct dependency on tika-parsers.  I'm testing a patch to address this.

mwoodiupui replied:
https://github.com/tdonohue/DSpace/pull/11 if you think it appropriate for this PR.

tdonohue replied:
Thanks @mwoodiupui : I've merged in your improvement to this PR...so solr-cell is now removed.

tdonohue committed:
Remove our custom Docker container for Solr. It's unreliable with Solr v8, and all the examples show to use docker-compose directly with official image.

tdonohue committed:
Upgrade Jetty to avoid security warnings from Snyk. Also fixes GitHub security alert

tdonohue commented:
@mwoodiupui : Thanks for your review! Just a quick note to let you know that I have addressed all your prior feedback. In the midst of that, I found a bug in the Solr v8 Docker scripts (now fixed) and did another Jetty minor update to resolve this security issue. https://github.com/DSpace/DSpace/security/dependabot/pom.xml/org.eclipse.jetty:jetty-server/open

If you have a chance to give it another look, I'd appreciate it! Thanks again

mwoodiupui reviewed (approved):
+1 by inspection.  This all looks reasonable to me.  (I did not give much attention to the Docker portions since I have so little contact with Docker.)

mwoodiupui commented on dspace-api/src/main/java/org/dspace/discovery/indexobject/IndexFactoryImpl.java:
```diff
@@ -74,35 +79,60 @@ public void writeDocument(Context context, T indexableObject, SolrInputDocument
+                ParseContext tikaContext = new ParseContext();
+
+                // Use Apache Tika to parse the full text stream
+                try {
```
Do we need a finally{} clause to close `streams`?

tdonohue replied:
Good catch...I think it is missing.  It's not `streams` we need to close, but `streams.getStream()` which creates a new `SequenceInputStream` that *does need to be closed*.  https://github.com/DSpace/DSpace/blob/main/dspace-api/src/main/java/org/dspace/discovery/FullTextContentStreams.java#L131

I had completely overlooked that, so I'm very glad you caught it!  I'll make sure this gets stored to a new variable & closed in a finally.

tdonohue replied:
Fixed now in https://github.com/DSpace/DSpace/pull/3020/commits/e5d5cc60e81cf19a9c2288ba6af3785b151cd1f9.  Thanks again @mwoodiupui !

tdonohue committed:
Ensure InputStream is closed

abollini reviewed (approved):
Hi @tdonohue thanks for this important work. Everything looks good to me so I'm +1 with merging it.
I would suggest to update to version 8.8.1 that is the latest stable and it seems perfectly compatible with what we already have (I have tested this PR against a local 8.8.1 installation without noting major issues). The switch to 8.8.1 can be done in the second PR #3126
